File: web/src/repo/data-catalog.ts
import { api } from './client'

// ─── Types ───────────────────────────────────────────────────────────────────

export interface DataDictionary {
  id: number
  catalog_id: number
  column_name: string
  data_type: string
  column_type?: string | null
  description: string | null
  is_nullable: boolean
  is_primary_key: boolean
  is_pii: boolean
  sample_values: string | null
  business_rule: string | null
  tags?: string | null
  created_at: string
  updated_at: string
}

export interface DataCatalog {
  id: number
  source_id: number | null
  destination_id: number | null
  table_name: string
  schema_name: string
  description: string | null
  owner: string | null
  tags: string | null
  classification: string | null
  sla_freshness_minutes: number | null
  row_count: number | null
  size_bytes: number | null
  last_analyzed_at: string | null
  created_at: string
  updated_at: string
  columns: DataDictionary[]
}

export interface DataCatalogCreate {
  source_id?: number | null
  destination_id?: number | null
  table_name: string
  schema_name?: string
  description?: string
  owner?: string
  tags?: string
  classification?: string
  row_count?: number
  size_bytes?: number
}

export interface DataCatalogUpdate {
  schema_name?: string
  table_name?: string
  description?: string
  owner?: string
  tags?: string
  classification?: string
  source_id?: number | null
  destination_id?: number | null
  row_count?: number
  size_bytes?: number
}

export interface DataCatalogListResponse {
  items: DataCatalog[]
  total: number
  page: number
  page_size: number
}

export interface DataDictionaryCreate {
  column_name: string
  data_type?: string
  description?: string
  is_nullable?: boolean
  is_primary_key?: boolean
  is_pii?: boolean
  sample_values?: string
  business_rule?: string
}

export interface DataDictionaryUpdate {
  column_name?: string
  data_type?: string
  description?: string
  is_nullable?: boolean
  is_primary_key?: boolean
  is_pii?: boolean
  sample_values?: string
  business_rule?: string
}

// ─── Repository ──────────────────────────────────────────────────────────────

export const dataCatalogRepo = {
  list(page = 1, pageSize = 20, search?: string) {
    return api.get<DataCatalogListResponse>('/data-catalog', {
      params: { page, page_size: pageSize, search },
    })
  },

  get(id: number) {
    return api.get<DataCatalog>(`/data-catalog/${id}`)
  },

  create(payload: DataCatalogCreate) {
    return api.post<DataCatalog>('/data-catalog', payload)
  },

  update(id: number, payload: DataCatalogUpdate) {
    return api.put<DataCatalog>(`/data-catalog/${id}`, payload)
  },

  remove(id: number) {
    return api.delete<{ message: string }>(`/data-catalog/${id}`)
  },

  // Dictionary columns
  getColumns(catalogId: number) {
    return api.get<DataDictionary[]>(`/data-catalog/${catalogId}/columns`)
  },

  addColumn(catalogId: number, payload: DataDictionaryCreate) {
    return api.post<DataDictionary>(`/data-catalog/${catalogId}/columns`, payload)
  },

  updateColumn(columnId: number, payload: DataDictionaryUpdate) {
    return api.put<DataDictionary>(`/data-catalog/columns/${columnId}`, payload)
  },

  removeColumn(columnId: number) {
    return api.delete<{ message: string }>(`/data-catalog/columns/${columnId}`)
  },
}

File: backend/app/domain/services/data_catalog.py
"""
Data Catalog service — business logic for data catalog & dictionary.
"""

from typing import List, Optional, Tuple

from sqlalchemy.orm import Session

from app.core.logging import get_logger
from app.domain.models.data_catalog import DataCatalog, DataDictionary
from app.domain.repositories.data_catalog import (
    DataCatalogRepository,
    DataDictionaryRepository,
)
from app.domain.schemas.data_catalog import (
    DataCatalogCreate,
    DataCatalogUpdate,
    DataDictionaryCreate,
    DataDictionaryUpdate,
)

logger = get_logger(__name__)


class DataCatalogService:
    """Business logic for Data Catalog management."""

    def __init__(self, db: Session):
        self.db = db
        self.catalog_repo = DataCatalogRepository(db)
        self.dictionary_repo = DataDictionaryRepository(db)

    # ─── Catalog CRUD ──────────────────────────────────────────────────────

    def create_catalog(self, data: DataCatalogCreate) -> DataCatalog:
        """Create a new catalog entry."""
        catalog = self.catalog_repo.create(
            source_id=data.source_id,
            destination_id=data.destination_id,
            table_name=data.table_name,
            schema_name=data.schema_name,
            description=data.description,
            owner=data.owner,
            classification=data.classification,
            sla_freshness_minutes=data.sla_freshness_minutes,
            tags=data.tags or [],
            custom_properties=data.custom_properties or {},
        )
        self.db.commit()
        self.db.refresh(catalog)
        logger.info(f"DataCatalog created: id={catalog.id} table={catalog.table_name}")
        return catalog

    def get_catalog(self, catalog_id: int) -> DataCatalog:
        """Get a catalog entry by ID."""
        return self.catalog_repo.get_by_id(catalog_id)

    def list_catalogs(
        self, skip: int = 0, limit: int = 20, search: str = None
    ) -> Tuple[List[DataCatalog], int]:
        """List catalogs with pagination and search."""
        return self.catalog_repo.get_all_paginated(
            skip=skip, limit=limit, search=search
        )

    def update_catalog(
        self, catalog_id: int, data: DataCatalogUpdate
    ) -> DataCatalog:
        """Update catalog metadata."""
        update_kwargs = data.dict(exclude_unset=True, exclude_none=True)
        if not update_kwargs:
            return self.catalog_repo.get_by_id(catalog_id)
        catalog = self.catalog_repo.update(catalog_id, **update_kwargs)
        self.db.commit()
        self.db.refresh(catalog)
        return catalog

    def delete_catalog(self, catalog_id: int) -> None:
        """Delete a catalog entry and its dictionary columns."""
        self.catalog_repo.delete(catalog_id)
        self.db.commit()
        logger.info(f"DataCatalog deleted: id={catalog_id}")

    # ─── Dictionary CRUD ──────────────────────────────────────────────────

    def list_columns(self, catalog_id: int) -> List[DataDictionary]:
        """List all dictionary columns for a catalog."""
        return self.dictionary_repo.get_by_catalog_id(catalog_id)

    def add_column(
        self, catalog_id: int, data: DataDictionaryCreate
    ) -> DataDictionary:
        """Add a column to the data dictionary."""
        # Verify catalog exists
        self.catalog_repo.get_by_id(catalog_id)
        col = self.dictionary_repo.create(
            catalog_id=catalog_id,
            column_name=data.column_name,
            data_type=data.data_type,
            description=data.description,
            is_pii=data.is_pii,
            is_nullable=data.is_nullable,
            sample_values=data.sample_values,
            business_rule=data.business_rule,
        )
        self.db.commit()
        self.db.refresh(col)
        return col

    def update_column(
        self, column_id: int, data: DataDictionaryUpdate
    ) -> DataDictionary:
        """Update a dictionary column."""
        update_kwargs = data.dict(exclude_unset=True, exclude_none=True)
        if not update_kwargs:
            return self.dictionary_repo.get_by_id(column_id)
        col = self.dictionary_repo.update(column_id, **update_kwargs)
        self.db.commit()
        self.db.refresh(col)
        return col

    def delete_column(self, column_id: int) -> None:
        """Delete a dictionary column."""
        self.dictionary_repo.delete(column_id)
        self.db.commit()

    def upsert_column(
        self, catalog_id: int, column_name: str, **kwargs
    ) -> DataDictionary:
        """Upsert a column by catalog + name."""
        col = self.dictionary_repo.upsert_column(
            catalog_id=catalog_id, column_name=column_name, **kwargs
        )
        self.db.commit()
        self.db.refresh(col)
        return col
